File: src/Reloaded.Memory.Buffers/BufferAllocator.Windows.cs
using System.Diagnostics;
using Reloaded.Memory.Buffers.Exceptions;
#if NET5_0_OR_GREATER
using System.Runtime.Versioning;
#endif
using Reloaded.Memory.Buffers.Native.Windows;
using Reloaded.Memory.Buffers.Structs;
using Reloaded.Memory.Buffers.Structs.Params;
using Reloaded.Memory.Buffers.Utilities;
using static Reloaded.Memory.Buffers.Native.Windows.Kernel32;
using static Reloaded.Memory.Buffers.Utilities.Mathematics;
using static Reloaded.Memory.Native.Windows.Kernel32;

namespace Reloaded.Memory.Buffers;

#pragma warning disable CA1416 // Validate platform compatibility
/// <summary>
/// Windows specific buffer allocator.
/// </summary>
public static partial class BufferAllocator
{
    // Devirtualized based on target.
    private static LocatorItem AllocateWindows(BufferAllocatorSettings settings) =>
        Polyfills.GetProcessId() == settings.TargetProcess.Id
        ? AllocateFast(new LocalKernel32(), GetMaxWindowsAddress(settings.TargetProcess), settings)
        : AllocateFast(new RemoteKernel32(settings.TargetProcess.Handle), GetMaxWindowsAddress(settings.TargetProcess), settings);

    private static unsafe LocatorItem AllocateFast<T>(T k32, nuint maxAddress, BufferAllocatorSettings settings) where T : IKernel32
    {
        maxAddress = Min(maxAddress, settings.MaxAddress);
        for (int x = 0; x < settings.RetryCount; x++)
        {
            // Until we get all of the pages.
            nuint currentAddress = settings.MinAddress;
            while (currentAddress <= maxAddress)
            {
                // Get our info from VirtualQueryEx.
                var memoryInformation = new MEMORY_BASIC_INFORMATION();
                var hasPage = k32.VirtualQuery(currentAddress, &memoryInformation);
                if (hasPage == 0)
                    break;

                // Add the page and increment address iterator to go to next page.
                if (TryAllocateBuffer(k32, ref memoryInformation, settings, out var item))
                    return item;

                currentAddress += memoryInformation.RegionSize;
            }
        }

        // See remarks on 'BruteForce' in BufferAllocatorSettings, as for why this code exists.
        // I'm not particularly fond of it, but what can you do?
        if (settings.BruteForce)
        {
            nuint currentAddress = settings.MinAddress;
            while (currentAddress <= maxAddress)
            {
                var memoryInformation = new MEMORY_BASIC_INFORMATION();
                var hasItem = k32.VirtualQuery(currentAddress, &memoryInformation);
                if (hasItem == 0)
                    break;

                if (TryAllocateBuffer(k32, ref memoryInformation, settings, out var item))
                    return item;

                currentAddress += (nuint)Cached.GetAllocationGranularity();
            }
        }

        throw new MemoryBufferAllocationException(settings.MinAddress, settings.MaxAddress, (int)settings.Size);
    }

    /// <summary>
    ///     Calculates the max memory address for a given process.
    /// </summary>
    /// <param name="process">The process in question.</param>
    internal static nuint GetMaxWindowsAddress(Process process)
    {
        if (Polyfills.GetProcessId() == process.Id)
        {
            // Note: In WOW64 mode, the following rules apply:
            // - If current process is large address aware, this will return 0xFFFEFFFF.
            // - If it is not LAA, this should return 0x7FFEFFFF.
            return Cached.GetMaxAddress();
        }

        // Is this Windows on Windows 64? (x86 app running on x64 Windows)
        bool hasIsWow64 = IsWow64Process(process.Handle, out var isWow64);
        GetSystemInfo(out SYSTEM_INFO systemInfo);

        nuint maxAddress = 0x7FFFFFFF; // 32bit max

        // If target is not using WoW64 emulation layer, trust GetSystemInfo for max address.
        if (hasIsWow64 && !isWow64)
            maxAddress = systemInfo.lpMaximumApplicationAddress;

        return maxAddress;
    }

    private static bool TryAllocateBuffer<T>(T k32, ref MEMORY_BASIC_INFORMATION pageInfo, BufferAllocatorSettings settings, out LocatorItem result) where T : IKernel32
    {
        result = default;
        // Fast return if page is not free.
        if (pageInfo.State != MEM_STATE.FREE)
            return false;

        Span<nuint> results = stackalloc nuint[4];
        foreach (var addr in GetBufferPointersInPageRange(ref pageInfo, (int)settings.Size, settings.MinAddress, settings.MaxAddress, results))
        {
            // ReSharper disable once RedundantCast
            nuint allocated = k32.VirtualAlloc(addr, (nuint)settings.Size);
            if (allocated == 0)
                continue;

            // Sanity test in case of '0' value input.
            if (allocated != addr)
            {
                k32.VirtualFree(allocated, (nuint)settings.Size);
                continue;
            }

            result = new LocatorItem(allocated, settings.Size);
            return true;
        }

        return false;
    }

    /// <summary>
    /// Checks if memory can be allocated inside <paramref name="pageInfo"/> provided the size, minimum and maximum pointer.
    /// </summary>
    /// <param name="pageInfo">Contains the information about a singular memory page.</param>
    /// <param name="bufferSize">The size that an allocation would occupy. Pre-aligned to page-size.</param>
    /// <param name="minimumPtr">The maximum pointer an allocation can occupy.</param>
    /// <param name="maximumPtr">The minimum pointer an allocation can occupy.</param>
    /// <param name="results">Span containing the results; must have at least 4 items.</param>
    /// <returns>Zero if the operation fails; otherwise positive value.</returns>
    private static Span<nuint> GetBufferPointersInPageRange(ref MEMORY_BASIC_INFORMATION pageInfo, int bufferSize, nuint minimumPtr,
        nuint maximumPtr, Span<nuint> results)
    {
        nuint pageStart = pageInfo.BaseAddress;
        nuint pageEnd = pageInfo.BaseAddress + pageInfo.RegionSize;
        int allocationGranularity = Cached.GetAllocationGranularity();
        return GetPossibleBufferAddresses(minimumPtr, maximumPtr, pageStart, pageEnd, (nuint) bufferSize, allocationGranularity, results);
    }

    private interface IKernel32
    {
        unsafe nuint VirtualQuery(nuint lpAddress, MEMORY_BASIC_INFORMATION* lpBuffer);
        nuint VirtualAlloc(UIntPtr lpAddress, UIntPtr dwSize);
        bool VirtualFree(UIntPtr lpAddress, UIntPtr dwSize);
    }

    private struct LocalKernel32 : IKernel32
    {
        public unsafe nuint VirtualQuery(nuint lpAddress, MEMORY_BASIC_INFORMATION* lpBuffer)
            => Kernel32.VirtualQuery(lpAddress, lpBuffer, (nuint) sizeof(MEMORY_BASIC_INFORMATION));

        public nuint VirtualAlloc(UIntPtr lpAddress, UIntPtr dwSize) =>
            Reloaded.Memory.Native.Windows.Kernel32.VirtualAlloc(lpAddress, dwSize, MEM_ALLOCATION_TYPE.MEM_RESERVE | MEM_ALLOCATION_TYPE.MEM_COMMIT, MEM_PROTECTION.PAGE_EXECUTE_READWRITE);

        public bool VirtualFree(UIntPtr lpAddress, UIntPtr dwSize) =>
            Reloaded.Memory.Native.Windows.Kernel32.VirtualFree(lpAddress, dwSize, MEM_ALLOCATION_TYPE.MEM_RELEASE);
    }

    private readonly struct RemoteKernel32 : IKernel32
    {
        private readonly nint _handle;
        public RemoteKernel32(nint handle) => _handle = handle;

        public unsafe nuint VirtualQuery(nuint lpAddress, MEMORY_BASIC_INFORMATION* lpBuffer)
            => Kernel32.VirtualQueryEx(_handle, lpAddress, lpBuffer, (nuint) sizeof(MEMORY_BASIC_INFORMATION));

        public nuint VirtualAlloc(UIntPtr lpAddress, UIntPtr dwSize) =>
            VirtualAllocEx(_handle, lpAddress, dwSize, MEM_ALLOCATION_TYPE.MEM_RESERVE | MEM_ALLOCATION_TYPE.MEM_COMMIT, MEM_PROTECTION.PAGE_EXECUTE_READWRITE);

        public bool VirtualFree(UIntPtr lpAddress, UIntPtr dwSize) =>
            Reloaded.Memory.Native.Windows.Kernel32.VirtualFreeEx(_handle, lpAddress, dwSize, MEM_ALLOCATION_TYPE.MEM_RELEASE);
    }
}

File: src-rust/src/internal/buffer_allocator_windows.rs
// Windows specific code for buffer allocator.
use crate::internal::buffer_allocator::get_possible_buffer_addresses;
use crate::structs::errors::BufferAllocationError;
use crate::structs::internal::LocatorItem;
use crate::structs::params::BufferAllocatorSettings;
use crate::utilities::cached::get_sys_info;
use crate::utilities::mathematics::min;
use crate::utilities::wrappers::Unaligned;
use core::ffi::c_void;
use core::mem::{size_of, zeroed};
use core::sync::atomic::AtomicI32;
use windows_sys::Win32::Foundation::{CloseHandle, BOOL, HANDLE};
use windows_sys::Win32::System::Memory::{
    VirtualAlloc, VirtualFree, VirtualQuery, MEMORY_BASIC_INFORMATION, MEM_COMMIT, MEM_FREE,
    MEM_RELEASE, MEM_RESERVE, PAGE_EXECUTE_READWRITE,
};
use windows_sys::Win32::System::SystemInformation::{GetSystemInfo, SYSTEM_INFO};
use windows_sys::Win32::System::Threading::IsWow64Process;
use windows_sys::Win32::System::Threading::{OpenProcess, PROCESS_ALL_ACCESS};

// Abstractions for kernel32 functions //
pub trait Kernel32 {
    fn virtual_query(
        &self,
        lp_address: *const c_void,
        lp_buffer: &mut MEMORY_BASIC_INFORMATION,
    ) -> usize;
    fn virtual_alloc(&self, lp_address: *const c_void, dw_size: usize) -> *mut c_void;
    fn virtual_free(&self, lp_address: *mut c_void, dw_size: usize) -> bool;
}

pub struct LocalKernel32;

impl Kernel32 for LocalKernel32 {
    fn virtual_query(
        &self,
        lp_address: *const c_void,
        lp_buffer: &mut MEMORY_BASIC_INFORMATION,
    ) -> usize {
        unsafe { VirtualQuery(lp_address, lp_buffer, size_of::<MEMORY_BASIC_INFORMATION>()) }
    }

    fn virtual_alloc(&self, lp_address: *const c_void, dw_size: usize) -> *mut c_void {
        unsafe {
            VirtualAlloc(
                lp_address,
                dw_size,
                MEM_RESERVE | MEM_COMMIT,
                PAGE_EXECUTE_READWRITE,
            )
        }
    }

    fn virtual_free(&self, lp_address: *mut c_void, dw_size: usize) -> bool {
        unsafe { VirtualFree(lp_address, dw_size, MEM_RELEASE) != 0 }
    }
}

#[cfg(feature = "external_processes")]
pub struct RemoteKernel32 {
    handle: HANDLE,
}

#[cfg(feature = "external_processes")]
impl Kernel32 for RemoteKernel32 {
    fn virtual_query(
        &self,
        lp_address: *const c_void,
        lp_buffer: &mut MEMORY_BASIC_INFORMATION,
    ) -> usize {
        unsafe {
            windows_sys::Win32::System::Memory::VirtualQueryEx(
                self.handle,
                lp_address,
                lp_buffer,
                size_of::<MEMORY_BASIC_INFORMATION>(),
            )
        }
    }

    fn virtual_alloc(&self, lp_address: *const c_void, dw_size: usize) -> *mut c_void {
        unsafe {
            windows_sys::Win32::System::Memory::VirtualAllocEx(
                self.handle,
                lp_address,
                dw_size,
                MEM_RESERVE | MEM_COMMIT,
                PAGE_EXECUTE_READWRITE,
            )
        }
    }

    fn virtual_free(&self, lp_address: *mut c_void, dw_size: usize) -> bool {
        unsafe {
            windows_sys::Win32::System::Memory::VirtualFreeEx(
                self.handle,
                lp_address,
                dw_size,
                MEM_RELEASE,
            ) != 0
        }
    }
}

// Abstractions for Process //
pub(crate) struct ProcessHandle {
    handle: HANDLE,
}

impl ProcessHandle {
    // open_process opens a new process with the given id and returns a ProcessHandle.
    pub unsafe fn open_process(id: u32) -> Self {
        ProcessHandle {
            handle: OpenProcess(PROCESS_ALL_ACCESS, 0, id),
        }
    }

    pub fn is_valid(&self) -> bool {
        self.handle != 0
    }

    #[cfg(feature = "external_processes")]
    pub fn get_raw_handle(&self) -> HANDLE {
        self.handle
    }
}

impl Drop for ProcessHandle {
    fn drop(&mut self) {
        unsafe {
            // Close non-null handle
            if self.handle != 0 {
                CloseHandle(self.handle);
            }
        }
    }
}

// Helpers //

fn get_max_windows_address(process_id: u32, handle: HANDLE) -> usize {
    if get_sys_info().this_process_id == process_id {
        // Note: In WOW64 mode, the following rules apply:
        // - If current process is large address aware, this will return 0xFFFEFFFF.
        // - If it is not LAA, this should return 0x7FFEFFFF.
        return get_sys_info().max_address;
    }

    unsafe {
        // Is this Windows on Windows 64? (x86 app running on x64 Windows)
        let mut is_wow64: BOOL = Default::default();
        let has_is_wow64 = IsWow64Process(handle, &mut is_wow64) != 0;

        let mut system_info: SYSTEM_INFO = zeroed();
        GetSystemInfo(&mut system_info);

        let mut max_address = 0x7FFFFFFF; // 32bit max

        // If target is not using WoW64 emulation layer, trust GetSystemInfo for max address.
        if is_wow64 == 0 && has_is_wow64 {
            max_address = system_info.lpMaximumApplicationAddress as usize;
        }

        max_address
    }
}

// Implementation //
pub fn allocate_windows(
    settings: &BufferAllocatorSettings,
) -> Result<LocatorItem, BufferAllocationError> {
    unsafe {
        let process_handle = ProcessHandle::open_process(settings.target_process_id);

        if !process_handle.is_valid() {
            return Err(BufferAllocationError::new(
                *settings,
                "Failed to open process",
            ));
        }

        let handle = process_handle.handle;
        let max_address = get_max_windows_address(settings.target_process_id, handle);

        #[cfg(feature = "external_processes")]
        {
            if get_sys_info().this_process_id == settings.target_process_id {
                allocate_fast(&LocalKernel32 {}, max_address, settings)
            } else {
                allocate_fast(&RemoteKernel32 { handle }, max_address, settings)
            }
        }

        #[cfg(not(feature = "external_processes"))]
        allocate_fast(&LocalKernel32 {}, max_address, settings)
    }
}

fn allocate_fast<T: Kernel32>(
    k32: &T,
    mut max_address: usize,
    settings: &BufferAllocatorSettings,
) -> Result<LocatorItem, BufferAllocationError> {
    max_address = min(max_address, settings.max_address);

    for _ in 0..settings.retry_count {
        // Until we get all of the pages.
        let mut current_address = settings.min_address;
        let mut memory_information: MEMORY_BASIC_INFORMATION = unsafe { zeroed() };
        while current_address <= max_address {
            // Get our info from VirtualQueryEx.
            let has_page =
                k32.virtual_query(current_address as *const c_void, &mut memory_information);

            if has_page == 0 {
                break;
            }

            // Add the page and increment address iterator to go to next page.
            match try_allocate_buffer(k32, &mut memory_information, settings) {
                Some(item) => return Ok(item),
                None => {
                    current_address += memory_information.RegionSize;
                }
            };
        }
    }

    // See remarks on 'BruteForce' in BufferAllocatorSettings, as for why this code exists.
    // I'm not particularly fond of it, but what can you do?
    if settings.brute_force {
        let mut current_address = settings.min_address;
        let mut memory_information = unsafe { zeroed() };
        while current_address <= max_address {
            let has_item =
                k32.virtual_query(current_address as *const c_void, &mut memory_information);
            if has_item == 0 {
                break;
            }

            match try_allocate_buffer(k32, &mut memory_information, settings) {
                Some(item) => return Ok(item),
                None => {
                    current_address += get_sys_info().allocation_granularity as usize;
                }
            };
        }
    }

    Err(BufferAllocationError::new(
        *settings,
        "Failed to allocate buffer on Windows",
    ))
}

fn try_allocate_buffer<T: Kernel32>(
    k32: &T,
    page_info: &mut MEMORY_BASIC_INFORMATION,
    settings: &BufferAllocatorSettings,
) -> Option<LocatorItem> {
    // Fast return if page is not free.
    if page_info.State != MEM_FREE {
        return None;
    }

    let mut results = [0; 4];
    for addr in get_buffer_pointers_in_page_range(
        page_info,
        settings.size as usize,
        settings.min_address,
        settings.max_address,
        &mut results,
    ) {
        let allocated = k32.virtual_alloc(*addr as *const c_void, settings.size as usize);

        if allocated.is_null() {
            continue;
        }

        // Sanity test just in case.
        if allocated as usize != *addr {
            k32.virtual_free(allocated, settings.size as usize);
            continue;
        }

        return Some(LocatorItem {
            base_address: Unaligned::new(allocated as usize),
            size: settings.size,
            position: 0,
            is_taken: AtomicI32::from(0),
        });
    }

    None
}

fn get_buffer_pointers_in_page_range<'a>(
    page_info: &mut MEMORY_BASIC_INFORMATION,
    buffer_size: usize,
    minimum_ptr: usize,
    maximum_ptr: usize,
    results: &'a mut [usize; 4],
) -> &'a [usize] {
    let page_start = page_info.BaseAddress as usize;
    let page_end = page_info.BaseAddress as usize + page_info.RegionSize;
    let allocation_granularity = get_sys_info().allocation_granularity;

    unsafe {
        get_possible_buffer_addresses(
            minimum_ptr,
            maximum_ptr,
            page_start,
            page_end,
            buffer_size,
            allocation_granularity as usize,
            results,
        )
    }
}
